### PR TITLE
manual: a cautionary internal runtime API section

### PR DESCRIPTION
**Pull request by Octachron**
This PR proposes to add a new "internal API" section to the manual. This section hopes to set some guidelines for people that, under duress or imperious necessity, stumble upon the internal runtime API. It is mostly a section of `don't but if you do` and of `really don't` that mostly documents the patch process established in #9205  .

### CODE REVIEW

File: manual/manual/cmds/intf-c.etex [hunk 1]
```diff
@@ -189,19 +189,15 @@ These files reside in the "caml/" subdirectory of the OCaml
...
-compatibility.
+{\bf Note:} It is highly recommended to define the macro "CAML_NAME_SPACE"
+on the command line with "-DCAML_NAME_SPACE" when including any of these
+header files. Otherwise, it is your responsability to ensure that
+the macro "CAML_NAME_SPACE" is defined bedore any of these header files.
+If you do not define it, the header files will also define short names
+(without the "caml_" prefix) for most functions, which usually produce
+clashes with names defined by other C libraries that you might use.
+Including the header files without "CAML_NAME_SPACE" is only supported
+for backward compatibility and may be removed in the future.
```
Comment (dbuenzli) [inline manual/manual/cmds/intf-c.etex]: I'm sorry I didn't follow the long discussions about this. What does this mean in practice ?

1. I should not do `ocamlc mystub.c` but `ocamlc -ccopt  '-DCAML_NAME_SPACE' mystub.c` ? Seems a bit annoying.
2. I can still `ocamlc mystub.c` but I should not complain if I use an unprefixed name and it no longer works in the future. 
3. ???

Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: So the paragraph is unclear, but the forward-compatible options are more:

1) use ocamlc -ccopt '-DCAML_NAME_SPACE' (I agree that this is a bit annoying)
2) define `CAML_nAME_SPACE` before importing any `caml/` headers

And the option that is only supported for backward compatibility and might disappear at some point:

3) use short names

Comment (dbuenzli) [reply manual/manual/cmds/intf-c.etex]: Note that in my 2. the idea was that at certain point `ocamlc file.c` would add the `-DCAML_NAME_SPACE` itself when short names are no longer supported.

I'm wondering if the instructions should then rather not be (and the sample code in the chapter be updated accordingly): before `#include`ing any `<caml/.*h>` you have to issue a `#define CAML_NAME_SPACE`. That way we can simply have people `ocamlc mystub.c` their c files.

The note should then simply mention that it is possible to do it without the #define but that this may bring clashes and short names and will not be supported in the future.

Comment (kayceesrk) [reply manual/manual/cmds/intf-c.etex]: I tend to prefer @dbuenzli suggestion. Suggesting macro definition is better than the explicit arguments since that lives with the code and hence, is explicit / tracked in source control, can be caught in code review, etc. With the compiler flag, I can imagine users may find it difficult to figure out how to get their high-level build tools (dune, ocamlbuild, etc) to pass the option down to the compiler. 

s/rather not be/rather be?

File: manual/manual/cmds/intf-c.etex [hunk 2]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+to the runtime internal API, this section provides some guidelines
+that may help you to write code that might not break on every new version
+of OCaml.
+
+\subsection{ss:c-internals}{CAML_INTERNALS}
+First, by default header files in the "caml/" subdirectory give only access
+to the public interface of the OCaml runtime.
+It is necessary to define the macro "CAML_INTERNALS" to get access to
+the lower-level interface. Once again, this lower-level interface is likely to
+change and break programs that use it.
```
Comment (kayceesrk) [inline manual/manual/cmds/intf-c.etex]: "s/likely to change/may change with every minor release" to make it explicit that there is really no guarantee.

File: manual/manual/cmds/intf-c.etex [hunk 3]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+which breaks in OCaml $\ge$ 4.10, you should include the "minor_gc" header:
+\begin{verbatim}
+#include <caml/minor_gc.h>
+\end{verbatim}
+
+\subsection{ss:c-internal-macros}{OCaml version macros}
+Finally, if including the right headers is not enough, or if you need to support
+version older than OCaml 4.04, the header file "caml/version.h" should help
+you to define your own compatibility layer.
+This file provides macros defining the current ocaml major
```
Comment (kayceesrk) [inline manual/manual/cmds/intf-c.etex]: s/ocaml/OCaml

File: manual/manual/cmds/intf-c.etex [hunk 4]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+#endif
+\end{verbatim}
+
+The "caml/version.h" header also defines a macro for the compiler variant
+"OCAML_VERSION_ADDITIONAL" which is used to distinguish alpha and beta releases
+or release candidates.
+Finally, the version string is available as "OCAML_VERSION_STRING".
+
+\subsection{ss:c-internal-no}{A final word of caution}
+Do not access the domain state directly. \\
```
Comment (kayceesrk) [inline manual/manual/cmds/intf-c.etex]: This section needs to be expanded to document what domain state is.

File: manual/manual/cmds/intf-c.etex (hunk 1 above)
Comment (gadmm) [reply manual/manual/cmds/intf-c.etex]: I do not really agree with the text. What might be deprecated is the usage of short names. The `CAML_NAME_SPACE` macro is a tool for programmers to transition and to ensure their code remains future-proof, it is not supposed to be a burden. I propose to change the text to say just that. I am highly sceptical that at some point OCaml will decide to throw an error if some header is included without CAML_NAME_SPACE.

File: manual/manual/cmds/intf-c.etex [hunk 5]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
 ("-L/usr/local/zlib") must be given on all three invocations of "ocamlmklib",
 because they are needed at different times depending on whether shared
 libraries are supported.
+
+
+\section{s:c-internal-guidelines}{Disavowed topic: the internal runtime API}
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: I would remove "Disavowed topic" (it is self-contradictory, and not really true because internals were already mentioned in the previous versions of the manual).

File: manual/manual/cmds/intf-c.etex [hunk 6]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+
+Let's start with a warning, there is \emph{no} stability guarantee if you are
+using internals of the OCaml runtime. If you really need access
+to the runtime internal API, this section provides some guidelines
+that may help you to write code that might not break on every new version
+of OCaml.
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex:2716]: I would turn this into a definition: internals are the bits in the `caml/` headers for which there is not stability guarantee.

File: manual/manual/cmds/intf-c.etex [hunk 7]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+
+\subsection{ss:c-internals}{CAML_INTERNALS}
+First, by default header files in the "caml/" subdirectory give only access
+to the public interface of the OCaml runtime.
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: This is not true: there is also the case of undocumented headers (e.g. `minor_gc.h`), which are considered internal but are not guarded with CAML_INTERNALS.

Comment (gadmm) [reply manual/manual/cmds/intf-c.etex]: Ideally I would change these to become guarded with CAML_INTERNALS, but that's a different patch (and I am not sure I have time for testing the change, etc.).

File: manual/manual/cmds/intf-c.etex [hunk 8]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+It is necessary to define the macro "CAML_INTERNALS" to get access to
+the lower-level interface. Once again, this lower-level interface is likely to
+change and break programs that use it.
+Another caveat is that this lower-level interface is only publicly accessible
+since OCaml 4.04.
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: This is not entirely true, some of it was already accessible before (and not guarded with CAML_INTERNALS IIRC). I would remove this sentence and reintroduce it in another form just below (see below).

File: manual/manual/cmds/intf-c.etex [hunk 9]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+
+\subsection{ss:c-internal-variables}{Internal variables}
+Second, if you are using internal C variables, do not redefine them
+by hand. You should import those variables by including the corresponding
+header files. The representation of those variables has already changed once in
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: I would rephrase these sentences positively in order to say that since 4.04, it is possible to import every internal variable by including some internal header files.

File: manual/manual/cmds/intf-c.etex [hunk 10]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+OCaml 4.10, and is still under evolution.
+If your code relies on such internal and brittle properties, it will be broken
+at some point in time.
+
+For instance, rather than redefining "caml_young_limit":
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex:2733]: I think this example is great to help people transition to 4.10, but I am not sure the manual is the best place for such a specific example. I'd really like if there was a centralised document that lists all the successive known transition actions like this (part of the OCaml or on some wiki).

File: manual/manual/cmds/intf-c.etex [hunk 11]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+...
+#else
+...
+#endif
+\end{verbatim}
+
+The "caml/version.h" header also defines a macro for the compiler variant
+"OCAML_VERSION_ADDITIONAL" which is used to distinguish alpha and beta releases
+or release candidates.
+Finally, the version string is available as "OCAML_VERSION_STRING".
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: This is not a full documentation: I do not know the format of the values contained in these macros. I think documenting `OCAML_VERSION` is enough (and it is great), and you can add that it is of the form `ABBCC`.

File: manual/manual/cmds/intf-c.etex [hunk 12]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+\subsection{ss:c-internal-macros}{OCaml version macros}
+Finally, if including the right headers is not enough, or if you need to support
+version older than OCaml 4.04, the header file "caml/version.h" should help
+you to define your own compatibility layer.
+This file provides macros defining the current ocaml major
+version "OCAML_VERSION_MAJOR", the minor version "OCAML_VERSION_MINOR", the
+patch level "OCAML_VERSION_PATCHLEVEL" and the full version
+"OCAML_VERSION" (which is the concatenation of the three previous macros).
+
+For example, if you need some specific handling for versions more recent than 4.10.0,
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: I'd write "older than 4.10.0" because if branch `>= 41000` is the one which is meant to remain, and the branch `< 41000` is the one included for backwards compatibility and subject to removal later.

File: manual/manual/cmds/intf-c.etex [hunk 13]
```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
...
+you could write
+\begin{verbatim}
+#include <caml/version.h>
+#if OCAML_VERSION > 41000
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: Realistically, that would more probably be `>= 41000`.

File: manual/manual/cmds/intf-c.etex [hunk 14]
```diff
@@ -189,19 +189,15 @@ These files reside in the "caml/" subdirectory of the OCaml
...
-before including these header files. If you do not define it, the
-header files will also define short names (without the "caml_" prefix)
-for most functions, which usually produce clashes with names defined
-by other C libraries that you might use. Including the header files
-without "CAML_NAME_SPACE" is only supported for backward
-compatibility.
+{\bf Note:} It is highly recommended to define the macro "CAML_NAME_SPACE"
+on the command line with "-DCAML_NAME_SPACE" when including any of these
+header files. Otherwise, it is your responsability to ensure that
+the macro "CAML_NAME_SPACE" is defined bedore any of these header files.
```
Comment (gadmm) [inline manual/manual/cmds/intf-c.etex]: typo at "before"

**Review by gadmm — CHANGES_REQUESTED** [review]
Thanks for the documentation. Overall you deal with all relevant subjects IIRC but there are a few things to fix. (edit: some are important to fix, some are advice)

**Comment by gasche** [comment]
A meta comment: I like code examples and I think that they help understand quickly how to use things (in addition to their precise definition in informal text). Having an example of best practice to include `.h` is great, and having relevant code examples on how to use the OCAML_VERSION stuff would be nice.

(And: thanks for doing the right thing and documenting these mechanisms -- which you didn't add in the first place. We have a lot of comments on informal text, but please feel free at any point to consider that what you have is "good enough" (it's certainly an improvement over what we had before), stop iterating, and ask for the current state to be merged.)

File: manual/manual/cmds/intf-c.etex (hunk 1 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: @dbuenzli , thanks for reminding me of updating the code examples!

Throwing an error might be too much, but we could start by printing a notification. I agree that it will probably better to introduce `CAML_NAME_SPACE` earlier and only mention the command line definition path as an alternative path.

File: manual/manual/cmds/intf-c.etex (hunk 7 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: In fact, it might be worth it to list all "considered" internal headers.

**Comment by gadmm** [comment]
About the Coq VM we complained that they had a serious use-case but never reported the shortcoming. I propose to add: `Programmers which come to rely on the internal API for a use-case which they find realistic and useful are encouraged to open a request for improvement on the bug tracker`.

File: manual/manual/cmds/intf-c.etex (hunk 4 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: Does it? It is not clear to me that we have information on the domain state that would be useful to the narrow group of expert users diving into the runtime internals that are not compiler hackers. It might be easier to just remove the section.

File: manual/manual/cmds/intf-c.etex (hunk 7 above)
Comment (gadmm) [reply manual/manual/cmds/intf-c.etex]: Ok but be careful of the forward-compatibility of the documentation, it becomes quickly outdated.

File: manual/manual/cmds/intf-c.etex (hunk 4 above)
Comment (gadmm) [reply manual/manual/cmds/intf-c.etex]: I posted the following comment earlier but it does not show up for me. I post it again:

I disagree with the advice to use `Caml_state_field`: it is only used if you want to be compatible with both short names and namespaced names. Ideally it is better if users do not use it at all. I'd rather advise people that they have to define CAML_NAME_SPACE and use `Caml_state->` directly.

However one should warn that it is internal, even if it can (currently) be accessed without CAML_INTERNALS. In the future, I even hope that it will be impossible to access Caml_state without it (#9179). Minimal documentation would count as partial fix for #9179

File: manual/manual/cmds/intf-c.etex (hunk 2 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: Good idea. Done.

File: manual/manual/cmds/intf-c.etex (hunk 4 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: As a simpler solution, I removed this section in the current version. We can add it later if needed.

File: manual/manual/cmds/intf-c.etex (hunk 5 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: This oxymoron might be a little too much indeed. I am proposing to settle for "Cautionary words".

File: manual/manual/cmds/intf-c.etex (hunk 6 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex:2716]: I am not sure how the reader is supposed to know which bits of caml/ offer no guarantee. I proposed a more negative definition: every header in "caml/" that is not described in the manual.

File: manual/manual/cmds/intf-c.etex (hunk 9 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: A more positive statement does flow better here, thanks!

File: manual/manual/cmds/intf-c.etex (hunk 10 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex:2733]: I think it is better to give a small local example for now and upgrade to a more centralized documentation when the need arises.

File: manual/manual/cmds/intf-c.etex (hunk 11 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: You are right, it is probably better to focus on OCAML_VERSION.

File: manual/manual/cmds/intf-c.etex (hunk 12 above)
Comment (Octachron) [reply manual/manual/cmds/intf-c.etex]: Good point.

**Review by gadmm — APPROVED** [review]
It all looks good to me, and I agree with your points.

This version decided not to mention `Caml_state`, which is fine (#9179 is still open to keep track of the issue).

**Review by gasche — APPROVED** [review]
I believe that the discussion on this PR has converged to a stable state. Whenever you feel like it, please add a Changes entry and then merge.

**Comment by Octachron** [comment]
Merged and cherry-picked to 4.10 as c87baeaf158772b2c2acd3 .
